src/logic/check.ts

```diff
@@ -0,0 +1,241 @@
+  const profileSource = await findLocalProfileSource(superJson, profile);
+  if (profileSource) {
+    profileAst = parseProfile(new Source(profileSource, profileId));
+    options?.logCb?.(`Profile: "${profileId}" found on local file system`);
+  } else {
+    //Load from store
+    options?.logCb?.(`Loading profile: "${profileId}" from Superface store`);
+    profileAst = await fetchProfileAST(profileId);
+    numberOfRemoteFilesUsed++;
+  }
+  if (!isProfileDocumentNode(profileAst)) {
+    throw userError(`Profile file has unknown structure`, 1);
+  }
+
+  //Load map AST
+  const mapSource = await findLocalMapSource(superJson, profile, provider);
+  if (mapSource) {
+    mapAst = parseMap(new Source(mapSource, `${profile.name}.${provider}`));
+    options?.logCb?.(
+      `Map for profile: "${profileId}" and provider: "${provider}" found on local filesystem`
+    );
+  } else {
+    //Load from store
+    options?.logCb?.(
+      `Loading map for profile: "${profileId}" and provider: "${provider}" from Superface store`
+    );
+    //TODO: use actual implementation
+    mapAst = await fetchMapAST(
+      profile.name,
+      provider,
+      profile.scope,
+      profile.version,
+      map.variant
+    );
+    numberOfRemoteFilesUsed++;
+  }
+
+  if (!isMapDocumentNode(mapAst)) {
+    throw userError(`Map file has unknown structure`, 1);
+  }
+
+  //Load provider.json
+  const localProviderJson = await findLocalProviderSource(superJson, provider);
+  if (localProviderJson) {
+    providerJson = localProviderJson;
+    options?.logCb?.(`Provider: "${provider}" found on local file system`);
+  } else {
+    options?.logCb?.(`Loading provider: "${provider}" from Superface store`);
+    providerJson = await fetchProviderInfo(provider);
+    numberOfRemoteFilesUsed++;
+  }
+
+  if (numberOfRemoteFilesUsed === 3) {
+    options?.warnCb?.(
+      `All files for specified capability have been downloaded - checking only remote files is redundant`
+    );
+  }
+
+  options?.logCb?.(
+    `Checking profile: "${profile.name}" and map for provider: "${provider}"`
+  );
+  //Check map and profile
+  const result = checkMapAndProfile(profileAst, mapAst, options);
+
+  options?.logCb?.(`Checking provider: "${provider}"`);
+  try {
+    parseProviderJson(providerJson);
+  } catch (error) {
+    //TODO: better way of formating?
+    if ('issues' in error) {
+      for (const issue of (error as { issues: [] }).issues) {
+        if ('path' in issue && 'message' in issue) {
+          result.push({
+            kind: 'error',
+            message: `Provider check error: ${
+              (issue as { message: string }).message
+            } on path ${(issue as { path: string }).path}`,
+          });
+        }
+      }
+    }
+  }
+
+  return result;
+}
+
+export function checkMapAndProfile(
+  profile: ProfileDocumentNode,
+  map: MapDocumentNode,
+  options?: {
+    strict?: boolean;
+    logCb?: LogCallback;
+  }
+): CheckResult[] {
```

Comment:
I thought this would use the linter/validator from parser?

Reply:
We use that in Lint I think :)

Reply:
Shouldn't we merge check and lint then? I believe the linter does the same checks as this code.

Reply:
I does similar stuff. Lint does not do:
Provider parsing (here we could maybe come up with some map/provider cross check, in the future)
Usecase matching

I am not against merging I just want to do maximum for ensuring that capability is ok.

Reply:
I'd eventually merge it all into the linter, but for now this is ok.